annotator-models/trainer/dawid_skene.py

```diff
@@ -0,0 +1,517 @@
+    for k in range(nRaters):
+        # [nClasses, nClasses] = [nClasses x nItems], [nItems, nClasses]
+        error_rates[k, :, :] = np.matmul(item_classes.T, counts[:,k,:])
+        sum_over_responses = np.sum(error_rates[k,:,:], axis=1)[:,None]
+
+        # divide each row by the sum over all observation classes
+        error_rates[k,:,:] = np.divide(
+            error_rates[k,:,:], sum_over_responses, where=sum_over_responses!=0)
+
+    return (class_marginals, error_rates)
+
+def e_step(counts, class_marginals, error_rates):
+    """
+    Determine the probability of each item belonging to each class,
+    given current ML estimates of the parameters from the M-step
+    See equation 2.5 in Dawid-Skene (1979)
+
+    Inputs:
+      counts: Array of how many times each rating was given
+          by each rater for each item
+      class_marginals: probability of a random item belonging to each class
+      error_rates: probability of rater k assigning a item in class j
+          to class l [raters, classes, classes]
+
+    Returns:
+      item_classes: Soft assignments of items to classes
+          [items x classes]
+    """
+    [nItems, nRaters, nClasses] = np.shape(counts)
+
+    item_classes = np.zeros([nItems, nClasses])
+
+    for i in range(nItems):
+
+        counts_i = np.stack([counts[i,:,:],counts[i,:,:]],axis=1)
+        estimate_1 = np.prod(np.power(error_rates,counts_i), axis=(0,2))
+        estimate_1 = class_marginals *  estimate_1
+        item_classes[i,:] = estimate_1
+
+        # normalize error rates by dividing by the sum over all classes
+        item_sum = np.sum(item_classes[i,:])
+        if item_sum > 0:
+            item_classes[i,:] = item_classes[i,:]/float(item_sum)
+
+    return item_classes
+
+def e_step_verbose(counts, class_marginals, error_rates):
+    """
+    Determine the probability of each item belonging to each class,
+    given current ML estimates of the parameters from the M-step
+    See equation 2.5 in Dawid-Skene (1979)
+
+    Inputs:
+      counts: Array of how many times each rating was given
+          by each rater for each item
+      class_marginals: probability of a random item belonging to each class
+      error_rates: probability of rater k assigning a item in class j
+          to class l [raters, classes, classes]
+
+    Returns:
+      item_classes: Soft assignments of items to classes
+          [items x classes]
+    """
+    [nItems, nRaters, nClasses] = np.shape(counts)
+
+    item_classes = np.zeros([nItems, nClasses])
+
+    for i in range(nItems):
+        for j in range(nClasses):
+            estimate = class_marginals[j]
+            estimate *= np.prod(np.power(error_rates[:,j,:], counts[i,:,:]))
+
+            item_classes[i,j] = estimate
+
+        # normalize error rates by dividing by the sum over all classes
+        item_sum = np.sum(item_classes[i,:])
+
+        if item_sum == 0 or math.isnan(item_sum):
```

Comment:
General comment about the division by zero / NaN checks: Not sure what the right was to do this is, but it seems like in most cases it would be a bug if the zero/NaN occurred - so if we are ignoring these cases we might not notice a problem. It could be that sometimes it does happen, which is not ideal but fine.. Ideally this never happens and we make these if statements into assertions